File: app/src/main/java/org/tasks/location/AndroidGeocoder.java
package org.tasks.location;

import static com.todoroo.andlib.utility.AndroidUtilities.assertNotMainThread;
import static org.tasks.data.Place.newPlace;

import android.content.Context;
import android.location.Address;
import java.io.IOException;
import java.util.List;
import org.tasks.data.Place;

@SuppressWarnings("unused")
public class AndroidGeocoder implements Geocoder {

  private final Context context;

  public AndroidGeocoder(Context context) {
    this.context = context;
  }

  @Override
  public Place reverseGeocode(MapPosition mapPosition) throws IOException {
    assertNotMainThread();

    android.location.Geocoder geocoder = new android.location.Geocoder(context);
    List<Address> addresses =
        geocoder.getFromLocation(mapPosition.getLatitude(), mapPosition.getLongitude(), 1);
    Place place = newPlace(mapPosition);
    if (addresses.isEmpty()) {
      return place;
    }

    Address address = addresses.get(0);
    if (address.getMaxAddressLineIndex() >= 0) {
      place.setName(address.getAddressLine(0));
      StringBuilder builder = new StringBuilder(place.getName());
      for (int i = 1; i <= address.getMaxAddressLineIndex(); i++) {
        builder.append(", ").append(address.getAddressLine(i));
      }
      place.setAddress(builder.toString());
    }
    if (address.hasLatitude() && address.hasLongitude()) {
      place.setLatitude(address.getLatitude());
      place.setLongitude(address.getLongitude());
    }
    place.setPhone(address.getPhone());
    place.setUrl(address.getUrl());
    return place;
  }
}

File: app/src/main/java/org/tasks/location/AndroidGeocoder.kt
package org.tasks.location

import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.tasks.data.Place
import org.tasks.data.Place.Companion.newPlace
import javax.inject.Inject

@Suppress("unused")
class AndroidGeocoder @Inject constructor(@ApplicationContext context: Context) : Geocoder {
    private val geocoder = if (android.location.Geocoder.isPresent()) {
        android.location.Geocoder(context)
    } else {
        null
    }

    override suspend fun reverseGeocode(mapPosition: MapPosition): Place =
            withContext(Dispatchers.IO) {
                val addresses = geocoder?.getFromLocation(mapPosition.latitude, mapPosition.longitude, 1) ?: emptyList()
                val place = newPlace(mapPosition)
                if (addresses.isEmpty()) {
                    return@withContext place
                }
                val address = addresses[0]
                if (address.maxAddressLineIndex >= 0) {
                    place.name = address.getAddressLine(0)
                    val builder = StringBuilder(place.name)
                    for (i in 1..address.maxAddressLineIndex) {
                        builder.append(", ").append(address.getAddressLine(i))
                    }
                    place.address = builder.toString()
                }
                if (address.hasLatitude() && address.hasLongitude()) {
                    place.latitude = address.latitude
                    place.longitude = address.longitude
                }
                place.phone = address.phone
                place.url = address.url
                place
            }
}